python/lsst/daf/butler/registry/obscore/_manager.py

```diff
@@ -0,0 +1,358 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = ["ObsCoreLiveTableManager"]
+
+import json
+import re
+import uuid
+from collections import defaultdict
+from collections.abc import Mapping
+from typing import TYPE_CHECKING, Dict, Iterable, List, Optional, Type, cast
+
+import sqlalchemy
+from lsst.daf.butler import (
+    Config,
+    DataCoordinate,
+    DataCoordinateIterable,
+    DatasetRef,
+    Dimension,
+    DimensionUniverse,
+)
+from lsst.sphgeom import Region
+from lsst.utils.iteration import chunk_iterable
+
+from ..interfaces import ObsCoreTableManager, VersionTuple
+from ._config import ConfigCollectionType, ObsCoreManagerConfig
+from ._records import ExposureRegionFactory, RecordFactory
+from ._schema import ObsCoreSchema
+
+if TYPE_CHECKING:
+    from ..interfaces import (
+        Database,
+        DatasetRecordStorageManager,
+        DimensionRecordStorageManager,
+        StaticTablesContext,
+    )
+
+_VERSION = VersionTuple(0, 0, 1)
+
+
+class _ExposureRegionFactory(ExposureRegionFactory):
+    """Find exposure region from a matching visit dimensions records."""
+
+    def __init__(self, dimensions: DimensionRecordStorageManager):
+        self.dimensions = dimensions
+        self.universe = dimensions.universe
+        self.exposure = self.universe["exposure"]
+        self.visit = self.universe["visit"]
+
+    def exposure_region(self, dataId: DataCoordinate) -> Optional[Region]:
+        # Docstring is inherited from a base class.
+        visit_definition_storage = self.dimensions.get(self.universe["visit_definition"])
+        if visit_definition_storage is None:
+            return None
+        exposureDataId = dataId.subset(self.exposure.graph)
+        records = visit_definition_storage.fetch(DataCoordinateIterable.fromScalar(exposureDataId))
+        # There may be more than one visit per exposure, they should nave the
+        # same  region, so we use arbitrary one.
+        record = next(iter(records), None)
+        if record is None:
+            return None
+        visit: int = record.visit
+
+        detector = cast(Dimension, self.universe["detector"])
+        if detector in dataId:
+            visit_detector_region_storage = self.dimensions.get(self.universe["visit_detector_region"])
+            if visit_detector_region_storage is None:
+                return None
+            visitDataId = DataCoordinate.standardize(
+                {
+                    "instrument": dataId["instrument"],
+                    "visit": visit,
+                    "detector": dataId["detector"],
+                },
+                universe=self.universe,
+            )
+            records = visit_detector_region_storage.fetch(DataCoordinateIterable.fromScalar(visitDataId))
```

Comment:
Actually, it looks like there's now a totally different approach to this in `dax_obscore`, and that's already able to handle lots of datasets at once.  Is this just dead code here?

Reply:
Would be nice indeed to have a vectorized `fetch`, I'll look at it when DM-31725 is done.

`dax_obscore` fetches too many records as an optimization in the assumption that it is doing bulk conversion and will need a large fraction of those records (which may not be true). That comes with a significant memory overhead which is probably fine for a standalone tool, but I would not want to do the same thing in a generic case. In the case of a regular ingest of raws (this is probably the only dataset that is going to use this logic) there won't be visit defined yet, so this should normally result in 0 records. In this case it will return early after trying to find a visit definition few lines above.